### PR TITLE
Update `ClusterEnvironment` docs

### LINKED ISSUE #7073: DDP for multi-node jobs in non-slurm environments
## 🐛 Bug

When running DDP with num_nodes > 1 without a SLURM environment, the global rank is not computed correctly: global rank ends up being always equal to local rank, disregarding number of nodes. My hunch for why this is happening is that num_processes is not computed correctly.

My current solution is to create a custom plugin and explicitly pass num_nodes, please see below. I would need a bit of guidance to see how we can fix the current pytorch-lightning code in a less hacky way.

The following code works using the latest dev version (1.3.0rc1)

```
...

trainer = pl.Trainer(
  gpus=hparams.gpus,
  max_epochs=hparams.max_epochs,
  num_nodes=hparams.num_nodes,
  distributed_backend='ddp' if hparams.gpus > 1 else None,
  sync_batchnorm=hparams.gpus > 1,
  precision=hparams.precision,
  fast_dev_run=hparams.fast_dev_run,
  plugins=CustomPlugin(num_nodes=hparams.num_nodes),
  logger=logger)

...
```

```
class CustomEnvironment(LightningEnvironment):
    def __init__(self, num_nodes):
        super().__init__()
        self._creates_children = num_nodes > 1

    def creates_children(self):
        return self._creates_children

    def global_rank(self):
        rank = None
        if 'RANK' in os.environ:
            rank = int(os.environ.get("RANK"))
        return rank

    def master_address(self):
        return os.environ.get("MASTER_ADDR", "127.0.0.1")

    def master_port(self):
        return int(os.environ.get("MASTER_PORT", super().master_port()))

    def world_size(self):
        world_size = None
        if 'WORLD_SIZE' in os.environ:
           world_size = int(os.environ.get('WORLD_SIZE'))
        return world_size

    def node_rank(self):
        return int(os.environ.get("NODE_RANK", super().node_rank()))

    def local_rank(self) -> int:
        return int(os.environ.get("LOCAL_RANK", super().local_rank()))


class CustomPlugin(DDPPlugin):
    def __init__(self, num_nodes, **kwargs: Union[Any, Dict[str, Any]]) -> None:
        self.num_nodes = num_nodes
        super().__init__(
            cluster_environment=CustomEnvironment(num_nodes),
            find_unused_parameters=False,
            **kwargs,
        )

    @property
    def distributed_sampler_kwargs(self):
        distributed_sampler_kwargs = dict(num_replicas=self.world_size, rank=self.global_rank)
        return distributed_sampler_kwargs

    def set_world_ranks(self):
        self.local_rank = self.cluster_environment.local_rank()
        self.node_rank = self.cluster_environment.node_rank()
        self.global_rank = self.cluster_environment.global_rank()
        self.world_size = self.cluster_environment.world_size()
        
        # local run, this falls back to the current logic which is broken for
        # ddp and num_nodes > 1, I suspect the problem is in self.num_processes
        if self.global_rank is None:
            if self.num_nodes > 1:
                raise ValueError("Expected num_nodes > 1.")
            self.global_rank = self.num_processes * self.node_rank + self.local_rank
        if self.world_size is None:
            if self.num_nodes > 1:
                raise ValueError("Expected num_nodes > 1.")
            self.world_size = self.num_nodes * self.num_processes
```

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Updates to cluster docs:
- Merged docs for general purpose cluster and SLURM cluster into one doc
- Added docs for Cluster Environment
- Updated Grid docs

Fixes #7073

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
LGTM

**Review by ananthsub — APPROVED** [review]

File: docs/source/clouds/cluster.rst [hunk 1]
```diff
@@ -198,11 +239,94 @@ See also the multi-node examples
...
+        dataset = MyDataset()
+        dist_sampler = torch.utils.data.distributed.DistributedSampler(dataset)
+        dataloader = Dataloader(dataset, sampler=dist_sampler)
+        return dataloader
+
+    # in your training script
+    trainer = Trainer(replace_sampler_ddp=False)
+
+
 Self-balancing architecture (COMING SOON)
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/clouds/cluster.rst:329]: I'd remove this as we've moved past this point with fully sharded
